### PR TITLE
more work towards performance improvements

### PR DESCRIPTION
**Pull request by tomerd**
motivation: remove redundant abstrations to make way for performance work

changes:
* get rid of BasePackageContainer and the inheritance
* refactor RepositoryPackageContainerConstraint
* adjust downstream code

no functional changes here, just code reorganization 

⚠️ based on https://github.com/apple/swift-package-manager/pull/3091, so let merge that one first

### CODE REVIEW

**Comment by tomerd** [comment]
@swift-ci please smoke test

File: Sources/PackageGraph/RepositoryPackageContainer.swift [hunk 1]
```diff
@@ -285,6 +285,14 @@ public class RepositoryPackageContainer: BasePackageContainer, CustomStringConve
             packageKind: identifier.kind,
             fileSystem: fs)
     }
+
+    public var isRemoteContainer: Bool? {
+        return true
+    }
+
+    public var description: String {
+        return "RepositoryPackageContainer(\(identifier.repository.url.debugDescription))"
```
Comment (yim-lee) [inline Sources/PackageGraph/RepositoryPackageContainer.swift:294]:
```suggestion
        return "PackageContainerConstraint(\(identifier.repository.url.debugDescription))"
```

**Review by yim-lee — COMMENTED** [review]
LGTM

**Review by abertelrud — APPROVED** [review]

**Comment by tomerd** [comment]
@swift-ci please smoke test

File: Sources/PackageGraph/RepositoryPackageContainer.swift (hunk 1 above)
Comment (tomerd) [reply Sources/PackageGraph/RepositoryPackageContainer.swift:294]: this one is actually correct

**Pull request merged**